unittests/BuildAnalysis/CriticalPathSolverTests.swift

```diff
@@ -0,0 +1,145 @@
+// This source file is part of the Swift.org open source project
+//
+// Copyright 2019 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See http://swift.org/LICENSE.txt for license information
+// See http://swift.org/CONTRIBUTORS.txt for Swift project authors
+
+import XCTest
+// The Swift package has llbuildSwift as module
+#if SWIFT_PACKAGE
+import llbuild
+import llbuildSwift
+#else
+import llbuild
+#endif
+import llbuildAnalysis
+
+class CriticalPathSolverTests: XCTestCase {
+    struct Input {
+        let keys: [BuildKey]
+        let lookup: [BuildKey: RuleResult]
+        let keyByName: [String: BuildKey]
+        
+        func path(with names: [String]) throws -> CriticalPath? {
+            let expectedKeys = names.compactMap { keyByName[$0] }
+            let rules = expectedKeys.compactMap { lookup[$0] }
+            guard expectedKeys.count == names.count, rules.count == names.count else { return nil }
+            return try CriticalPath.construct(Array(zip(expectedKeys, rules)))
+        }
+    }
+    
+    typealias InputData = [(key: String, deps: [String], start: Double, end: Double)]
+    static let exampleData: InputData = [
+        ("A", [], 0, 1),
+        ("B", ["A"], 1, 2),
+        ("C", [], 1, 2),
+        ("D", ["B"], 2, 3),
+        ("E", [], 2.1, 3.1),
+        ("F", ["D", "E"], 3.5, 4.5),
+    ]
+    
+    func createInput(data: InputData = CriticalPathSolverTests.exampleData) -> Input {
+        let keys = data.map { BuildKey.Command(name: $0.key) }
+        let keyByName: [String: BuildKey] = Dictionary(uniqueKeysWithValues: keys.map { ($0.name, $0) })
+        let ruleResults: [RuleResult] = data.map {
+            let deps = $0.deps.map { BuildKey.Command(name: $0) }
+            let result = RuleResult(value: BuildValue.DirectoryTreeSignature(signature: 0x1234), signature: 0x1234, computedAt: 1, builtAt: 2, start: $0.start, end: $0.end, dependencies: deps)
+            return result
+        }
+        return Input(keys: keys, lookup: [BuildKey: RuleResult](zip(keys, ruleResults), uniquingKeysWith: {a, b in a}), keyByName: keyByName)
+    }
+    
+    func createSolver(input: Input) -> CriticalPathSolver {
+        return CriticalPathSolver(keys: input.keys, ruleResultLookup: { input.lookup[$0] })
+    }
+    
+    func testCriticalPath_emptyKeys() throws {
```

Comment:
The class name indicates these are critical path tests, no need to duplicate that